### ISSUE TITLE
add a basic implementation of spherical k-means

### ISSUE THREAD

**Issue by asmahani** [asmahani/TextEmbeddingFE#4]
we can start off by Lloyd's algorithm and simply switch out Euclidean distance and averaging logic with cosine distance and averaging logic.

**Comment by asmahani** [comment]
added an implementation of spherical k-means, but i'm getting an error that suggests division by zero occurs while calculating cosine distances. this is most likely due to initilization of centroids using one of the training points, which means L2 distance is zero.

**Comment by asmahani** [comment]
In an attempt to be more systematic about this feature development, I created a new branch. Also, below is my initial list of features to be included/excluded in the first version of spherical k-means:

### Included

- multiple random starts (and picking the one with smallest 'inertia')
- parallelization (of multiple starts)
- sample weights (?)
- Lloyd algorithm

### Excluded

- sparse data
- C/C++ implementation
- Elkan algorithm
- Online algorithm (mini-batches)
- Initialization methods other than 'random'

**Comment by asmahani** [comment]
we now have a decent implementation, but still slow and needs a C backend. pure Cython version of lloyd iteration didn't help much. wil revisit this later. also, didn't add parallelization since in current use case this is wrapped in grid search and other loops where the parllelization is more appropriate. also, we handled the error, which was due to some centroids ending up with zero members. this itself was driven primarily by starting centroids being identical samples from input rows. we first applied a unique to X before drawing samples from it, and second we have a try/catch block that restarts the algorithm if empty clusters are deteced. good enough for now ...
